cmd/podman/images/scp.go

```diff
@@ -0,0 +1,344 @@
+// save or load the image to. returns a string with the file name or an error
+func makeRemoteFile(dial *ssh.Client) (string, error) {
+	run := "mktemp"
+	remoteFile, err := connection.ExecRemoteCommand(dial, run)
+	if err != nil {
+		return "", err
+	}
+	remoteFile = strings.TrimSuffix(remoteFile, "\n")
+	if err != nil {
+		return "", err
+	}
+	return remoteFile, nil
+}
+
+// createConnections takes a boolean determining which ssh client to dial
+// and returns the dials client, its newly opened remote file, and an error if applicable.
+func createConnection(fromOrToRemote bool) (*ssh.Client, string, error) {
+	ind := 0
+	if !fromOrToRemote && len(scpOpts.Connections) > 1 {
+		ind = 1
+	}
+	cfg, err := connection.ValidateAndConfigure(scpOpts.URI[ind], scpOpts.Iden[ind])
+	if err != nil {
+		return nil, "", err
+	}
+	dialAdd, err := ssh.Dial("tcp", scpOpts.URI[ind].Host, cfg) // dial the client
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "failed to connect")
+	}
+	file, err := makeRemoteFile(dialAdd)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return dialAdd, file, nil
+}
+
+// validateImageName makes sure that the image given is valid and no injections are occurring
+// we simply use this for error checking, bot setting the image
+func validateImageName(input string) error {
+	// ParseNormalizedNamed transforms a shortname image into its
+	// full name reference so busybox => docker.io/library/busybox
+	// we want to keep our shortnames, so only return an error if
+	// we cannot parse what th euser has given us
+	_, err := reference.ParseNormalizedNamed(input)
+	return err
+}
+
+// parseArgs returns the valid connection data based off of the information provided by the user
+// args is an array of the command arguments and cfg is tooling configuration used to get service destinations
+// returned is serv and an error if applicable. serv is a map of service destinations with the connection name as the index
+// this connection name is intended to be used as EngineConfig.ServiceDestinations
+// this function modifies the global scpOpt entities: FromRemote, ToRemote, Connections, and SourceImageName
+func parseArgs(args []string, cfg *config.Config) (map[string]config.Destination, error) {
+	iden := []string{}
+	url := []string{}
+	serv := map[string]config.Destination{}
+	cliConnections := []string{}
+	switch len(args) {
+	case 1:
+		if strings.Contains(args[0], "::") {
+			scpOpts.FromRemote = true
+			cliConnections = append(cliConnections, args[0])
+		} else {
+			err := validateImageName(args[0])
+			if err != nil {
+				return nil, err
+			}
+			scpOpts.SourceImageName = args[0]
+		}
+	case 2:
+		if strings.Contains(args[0], "::") {
+			if !(strings.Contains(args[1], "::")) && len(strings.Split(args[0], "::")[1]) <= 1 { // if an image is specified, this mean we are loading from our client
+				cliConnections = append(cliConnections, args[0])
+				scpOpts.ToRemote = true
```

Comment:
Isn’t this `image scp foo:: bar`? Why is that `ToRemote`? What sets `SourceImageName`?

Reply:
similar to [this](https://github.com/containers/podman/pull/10828#discussion_r680100814l) this ensures that both `image scp foo:: bar` is the same as `image scp bar foo::` order of arguments shouldn't make a difference if they are valid

Reply:
added a setting for `SourceImageName` I had never tested that path.

Reply:
> similar to [this](https://github.com/containers/podman/pull/10828#discussion_r680100814l) this ensures that both `image scp foo:: bar` is the same as `image scp bar foo::` order of arguments shouldn't make a difference if they are valid

I don’t understand. How can the order of arguments not make a difference? One of the two means “copy remote file to local”, and the other to “copy local to remote”, doesn’t it? At least that I would expect based on the more generic case `image scp host1::image1 host2::image2`.

Or is this designed so that `image scp foo::bar` is remote→local, and a two-argument version is local→remote? If so, what will be the future syntax for remote→local while saving under a different name?

(The man page does not explain the meaning of the syntax either; it just shows the syntax.)

Reply:
> added a setting for SourceImageName I had never tested that path.

(That’s ultimately what unit tests are for, but that might truly be a later PR.)